src/Workspaces/Core/Portable/Workspace/Solution/Checksum_Factory.cs

```diff
@@ -91,13 +91,44 @@ public static Checksum Create(ArrayBuilder<Checksum> checksums)
                 checksum.WriteTo(writer);
         });
 
+    public static Checksum CreateNew(ArrayBuilder<Checksum> checksums)
+    {
+        var checksumsCount = checksums.Count;
+        if (checksumsCount <= 100)
+        {
+            Span<Checksum> hashes = stackalloc Checksum[checksumsCount];
+            for (var i = 0; i < checksumsCount; i++)
+                hashes[i] = checksums[i];
+
+            return Create(hashes);
+        }
+
+        return Create(checksums);
+    }
+
+    public static Checksum CreateNew(ImmutableArray<Checksum> checksums)
+    {
+        var hashes = ImmutableCollectionsMarshal.AsArray(checksums).AsSpan();
+
+        return Create(hashes);
+    }
+
     public static Checksum Create(ImmutableArray<Checksum> checksums)
         => Create(checksums, static (checksums, writer) =>
         {
             foreach (var checksum in checksums)
                 checksum.WriteTo(writer);
         });
 
+    public static Checksum CreateNew(ImmutableArray<byte> bytes)
```

Comment:
one nagging issue i have with this is ensuring we don't end up in collision situations.  So, like getting the checksum for an empty array of bytes, will produced the same checksum as an empty enumerable of strings.  I don't know if that's a concern or not.  And it's likely a problem external to this PR.